pkg/deploytest/simulation.go

```diff
@@ -0,0 +1,269 @@
+	moduleChan := testsim.NewChan()
+	n.moduleChans[id] = moduleChan
+
+	switch m := m.(type) {
+	case modules.ActiveModule:
+		return n.wrapActive(m, moduleChan)
+	case modules.PassiveModule:
+		return n.wrapPassive(m, moduleChan)
+	default:
+		panic("Unexpected module type")
+	}
+}
+
+// Start initiates simulation with the events from the write-ahead log
+// on behalf of the given process. To be called concurrently with
+// mir.Node.Run().
+func (n *SimNode) Start(proc *testsim.Process, walEvents *events.EventList) {
+	n.SendEvents(proc, walEvents)
+	for m := range n.moduleChans {
+		initEvents := events.EmptyList()
+		initEvents.PushBack(events.Init(m))
+		n.SendEvents(proc, initEvents)
+	}
+}
+
+type applyEventsFn func(ctx context.Context, eventList *events.EventList) (*events.EventList, error)
+
+type eventsIn struct {
+	ctx       context.Context
+	eventList *events.EventList
+}
+
+type eventsOut struct {
+	eventList *events.EventList
+	err       error
+}
+
+type simModule struct {
+	*SimNode
+	applyFn applyEventsFn
+	inChan  chan eventsIn
+	outChan chan eventsOut
+	simChan *testsim.Chan
+}
+
+func newSimModule(n *SimNode, m modules.Module, simChan *testsim.Chan) *simModule {
+	var applyFn applyEventsFn
+	switch m := m.(type) {
+	case modules.PassiveModule:
+		applyFn = func(_ context.Context, eventList *events.EventList) (*events.EventList, error) {
+			return m.ApplyEvents(eventList)
+		}
+	case modules.ActiveModule:
+		applyFn = func(ctx context.Context, eventList *events.EventList) (*events.EventList, error) {
+			return events.EmptyList(), m.ApplyEvents(ctx, eventList)
+		}
+	default:
+		panic("Unexpected module type")
+	}
+
+	sm := &simModule{
+		SimNode: n,
+		applyFn: applyFn,
+		inChan:  make(chan eventsIn, 1),
+		outChan: make(chan eventsOut, 1),
+		simChan: simChan,
+	}
+
+	go sm.run(n.Spawn(), applyFn)
+
+	return sm
+}
+
+func (m *simModule) run(proc *testsim.Process, applyFn applyEventsFn) {
+	origEvents := events.EmptyList()
+	for {
+		if origEvents.Len() == 0 {
+			newOrigEvents, ok := m.SimNode.recvEvents(proc, m.simChan)
+			if !ok {
+				return
+			}
+			origEvents.PushBackList(newOrigEvents)
+		}
+
+		in := <-m.inChan
+
+		for origEvents.Len() < in.eventList.Len() {
+			newOrigEvents, ok := m.SimNode.recvEvents(proc, m.simChan)
+			if !ok {
+				return
+			}
+			origEvents.PushBackList(newOrigEvents)
+		}
+
+		it := in.eventList.Iterator()
+		for e := it.Next(); e != nil; e = it.Next() {
+			if !proc.Delay(m.SimNode.delayFn(e)) {
+				return
+			}
+		}
+
+		var out eventsOut
+		out.eventList, out.err = applyFn(in.ctx, in.eventList)
+
+		if out.err == nil {
+			it := origEvents.Iterator()
+			followUps := events.EmptyList()
+			for i := 0; i < in.eventList.Len(); i++ {
+				followUps.PushBackSlice(it.Next().Next)
+			}
+			origEvents = events.EmptyList()
+			for e := it.Next(); e != nil; e = it.Next() {
+				origEvents.PushBack(e)
+			}
+			followUps.PushBackList(out.eventList)
```

Comment:
Honestly, I often find the event lists quite inconvenient to use, but I suppose there was some rational to introduce them instead of simply using slices. I believe the usage of the iterator here is correct. The first loop goes over as many leading items in `origEvents` as there are events passed to `ApplyEvents` (`in.eventList`). For each of those items, it collects the attached follow-up events in the `followUps` event list. The second loop continues iterating through the rest of `origEvents` and appends the remaining there events to a new empty event list reassigned to `origEvents`. If the event lists were slices, it would probably look like follows:

```go
var followUps []*eventpb.Event
for range in.eventList {
        followUps = append(followUps, origEvents[0].Next...)
        origEvents[0], origEvents = nil, origEvents[1:]
}
followUps = append(followUps, out.eventList...)
```